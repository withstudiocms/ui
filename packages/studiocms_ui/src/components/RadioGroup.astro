---
import type { StudioCMSColorway } from '../utils/colors';
import { generateID } from '../utils/generateID';

/**
 * The props for the RadioGroup component.
 */
interface Option {
  /**
   * The label of the option.
   */
	label: string;
  /**
   * The value of the option.
   */
	value: string;
  /**
   * Whether the option is disabled.
   */
	disabled?: boolean;
}

/**
 * The props for the RadioGroup component.
 */
interface Props {
  /**
   * The label of the radio group.
   */
	label: string;
  /**
   * The color of the radio group. Defaults to `default`.
   */
	color?: StudioCMSColorway;
  /**
   * The default value of the radio group. Needs to be one of the values in the options.
   */
	defaultValue?: string;
  /**
   * The options to display in the radio group.
   */
	options: Option[];
  /**
   * Whether the radio group is disabled. Defaults to `false`.
   */
	disabled?: boolean;
  /**
   * The name of the radio group.
   */
	name?: string;
  /**
   * Whether the radio group is required. Defaults to `false`.
   */
	isRequired?: boolean;
  /**
   * Whether the radio group is horizontal. Defaults to `false`.
   */
	horizontal?: boolean;
  /**
   * Additional classes to apply to the radio group.
   */
	class?: string;
}

const {
	label,
	color = 'default',
	defaultValue,
	options,
	disabled = false,
	isRequired = false,
	horizontal = false,
	name = generateID('radio'),
	class: className,
} = Astro.props;
---

<div 
  class="sui-radio-container"
  class:list={[
    disabled && "disabled",
    horizontal && "horizontal",
    color,
    className
  ]}
>
  <span>
    {label} <span class="req-star">{isRequired && "*"}</span>
  </span>
  <div class="sui-radio-inputs" role="radiogroup">
    {options.map(({ label, value, disabled: individuallyDisabled }, i) => (
      <label
        for={value}
        class="sui-radio-label"
        class:list={[ individuallyDisabled && "disabled" ]}
      >
        <div class="sui-radio-box-container">
          <div 
            class="sui-radio-box" 
            role="radio" 
            tabindex={i === 0 ? 0 : -1} 
            aria-checked={value === defaultValue}
            aria-label={label}
          />
        </div>
        <input 
          class="sui-radio-toggle"
          type="radio" 
          value={value} 
          id={value} 
          name={name} 
          checked={value === defaultValue} 
          disabled={disabled || individuallyDisabled}
          required={isRequired}
        />
        <span>{label}</span>
      </label>
    ))}
  </div>
</div>
<script>
  const elements = document.querySelectorAll<HTMLDivElement>('.sui-radio-container');

  for (const element of elements) {
    if (element.dataset.initialized) continue;
    
    element.dataset.initialized = 'true';

    const radioBoxes = element.querySelectorAll<HTMLDivElement>('.sui-radio-box');

    let i = 0;

    for (const radioBox of radioBoxes) {
      radioBox.addEventListener('keydown', (e) => {
        if (e.key === 'Enter' || e.key === " ") {
          e.preventDefault();

          const input = (e.target as HTMLDivElement).parentElement!.parentElement!.querySelector<HTMLInputElement>('.sui-radio-toggle')!;

          if (input.disabled) return;

          input.checked = true;
        }

        if (e.key === 'ArrowRight' || e.key === 'ArrowDown') {
          e.preventDefault();

          let nextRadioBox: HTMLDivElement | undefined;

          radioBoxes.forEach((box, index) => {
            if (box === radioBox) nextRadioBox = radioBoxes[index + 1];
          });

          if (!nextRadioBox) return;

          radioBox.tabIndex = -1;
          nextRadioBox.tabIndex = 0;
          nextRadioBox.focus();
          nextRadioBox.click();
        }

        if (e.key === 'ArrowLeft' || e.key === 'ArrowUp') {
          e.preventDefault();

          let previousRadioBox: HTMLDivElement | undefined;

          radioBoxes.forEach((box, index) => {
            if (box === radioBox) previousRadioBox = radioBoxes[index - 1];
          });

          if (!previousRadioBox) return;

          radioBox.tabIndex = -1;
          previousRadioBox.tabIndex = 0;
          previousRadioBox.focus();
          previousRadioBox.click();
        }
      });

      i++;
    }
    element.addEventListener('keydown', (e) => {
      if (e.key !== 'Enter') return;

      const checkbox = element.querySelector<HTMLInputElement>('.sui-checkbox');
        
      if (!checkbox) return;

      checkbox.click();
    });
  }
</script>
<style>
  .sui-radio-container {
    display: flex;
    flex-direction: column;
    gap: .5rem;
  }

  .sui-radio-container.disabled {
    opacity: 0.5;
    color: hsl(var(--text-muted));
  }

  .sui-radio-label.disabled {
    opacity: 0.5;
    color: hsl(var(--text-muted));
    pointer-events: none;
  }

  .req-star {
    color: hsl(var(--danger-base));
    font-weight: 700;
  }

  .sui-radio-inputs {
    display: flex;
    flex-direction: column;
    gap: .75rem;
  }

  .sui-radio-container.horizontal .sui-radio-inputs {
    flex-direction: row;
  }

  .sui-radio-label {
    display: flex;
    flex-direction: row;
    gap: .5rem;
    position: relative;
    align-items: center;
  }

  .sui-radio-label:hover .sui-radio-box {
    outline-color: hsl(var(--default-hover));
  }

  .sui-radio-container:not(.disabled) .sui-radio-label:active .sui-radio-box {
    outline-color: hsl(var(--default-active));
    scale: 0.9;
  }

  .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
    background-color: hsl(var(--text-normal));
    outline-color: hsl(var(--text-normal));
  }

  .sui-radio-container.primary .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
    background-color: hsl(var(--primary-base));
    outline-color: hsl(var(--primary-base));
  }

  .sui-radio-container.success .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
    background-color: hsl(var(--success-base));
    outline-color: hsl(var(--success-base));
  }

  .sui-radio-container.warning .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
    background-color: hsl(var(--warning-base));
    outline-color: hsl(var(--warning-base));
  }

  .sui-radio-container.danger .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
    background-color: hsl(var(--danger-base));
    outline-color: hsl(var(--danger-base));
  }

  .sui-radio-box-container {
    width: 20px;
    height: 20px;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    cursor: pointer;
  }

  .sui-radio-box {
    width: 12px;
    height: 12px;
    border-radius: 20px;
    outline: 1px solid hsl(var(--default-base));
    outline-offset: 4px;
    transition: all .15s ease;
  }

  .sui-radio-box:focus-visible {
    outline-color: hsl(var(--text-normal)) !important;
  }

  .sui-radio-toggle {
    width: 0;
    height: 0;
    visibility: hidden;
    opacity: 0;
    margin: 0;
    position: absolute;
  }
</style>
