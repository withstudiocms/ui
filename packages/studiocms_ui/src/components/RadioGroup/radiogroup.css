.sui-radio-container {
  display: flex;
  flex-direction: column;
  gap: .5rem;
}

.sui-radio-container.disabled {
  opacity: 0.5;
  color: hsl(var(--text-muted));
}

.sui-radio-label.disabled {
  opacity: 0.5;
  color: hsl(var(--text-muted));
  pointer-events: none;
}

.req-star {
  color: hsl(var(--danger-base));
  font-weight: 700;
}

.sui-radio-inputs {
  display: flex;
  flex-direction: column;
  gap: .75rem;
}

.sui-radio-container.horizontal .sui-radio-inputs {
  flex-direction: row;
}

.sui-radio-label {
  display: flex;
  flex-direction: row;
  gap: .5rem;
  position: relative;
  align-items: center;
}

.sui-radio-label:hover .sui-radio-box {
  outline-color: hsl(var(--default-hover));
}

.sui-radio-container:not(.disabled) .sui-radio-label:active .sui-radio-box {
  outline-color: hsl(var(--default-active));
  scale: 0.9;
}

.sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
  background-color: hsl(var(--text-normal));
  outline-color: hsl(var(--text-normal));
}

.sui-radio-container.primary .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
  background-color: hsl(var(--primary-base));
  outline-color: hsl(var(--primary-base));
}

.sui-radio-container.success .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
  background-color: hsl(var(--success-base));
  outline-color: hsl(var(--success-base));
}

.sui-radio-container.warning .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
  background-color: hsl(var(--warning-base));
  outline-color: hsl(var(--warning-base));
}

.sui-radio-container.danger .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
  background-color: hsl(var(--danger-base));
  outline-color: hsl(var(--danger-base));
}

.sui-radio-container.info .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
  background-color: hsl(var(--info-base));
  outline-color: hsl(var(--info-base));
}

.sui-radio-container.mono .sui-radio-label:has(.sui-radio-toggle:checked) .sui-radio-box {
  background-color: hsl(var(--mono-base));
  outline-color: hsl(var(--mono-base));
}

.sui-radio-box-container {
  width: 20px;
  height: 20px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  cursor: pointer;
}

.sui-radio-box {
  width: 12px;
  height: 12px;
  border-radius: var(--radius-full);
  outline: 1px solid hsl(var(--default-base));
  outline-offset: 4px;
  transition: all .15s ease;
}

.sui-radio-box:focus-visible {
  outline-color: hsl(var(--text-normal)) !important;
}

.sui-radio-toggle {
  width: 0;
  height: 0;
  visibility: hidden;
  opacity: 0;
  margin: 0;
  position: absolute;
}
