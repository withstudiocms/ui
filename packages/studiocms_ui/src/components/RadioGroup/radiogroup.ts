function loadRadioGroups() {
	const AllRadioGroupContainers = document.querySelectorAll<HTMLDivElement>('.sui-radio-container');

	for (const element of AllRadioGroupContainers) {
		if (element.dataset.initialized) continue;

		element.dataset.initialized = 'true';

		const radioBoxes = element.querySelectorAll<HTMLDivElement>('.sui-radio-box');

		let i = 0;

		for (const radioBox of radioBoxes) {
			radioBox.addEventListener('keydown', (e) => {
				if (e.key === 'Enter' || e.key === ' ') {
					e.preventDefault();

					const input = (
						e.target as HTMLDivElement
					).parentElement!.parentElement!.querySelector<HTMLInputElement>('.sui-radio-toggle')!;

					if (input.disabled) return;

					input.checked = true;
				}

				if (e.key === 'ArrowRight' || e.key === 'ArrowDown') {
					e.preventDefault();

					let nextRadioBox: HTMLDivElement | undefined;

					radioBoxes.forEach((box, index) => {
						if (box === radioBox) nextRadioBox = radioBoxes[index + 1];
					});

					if (!nextRadioBox) return;

					radioBox.tabIndex = -1;
					nextRadioBox.tabIndex = 0;
					nextRadioBox.focus();
					nextRadioBox.click();
				}

				if (e.key === 'ArrowLeft' || e.key === 'ArrowUp') {
					e.preventDefault();

					let previousRadioBox: HTMLDivElement | undefined;

					radioBoxes.forEach((box, index) => {
						if (box === radioBox) previousRadioBox = radioBoxes[index - 1];
					});

					if (!previousRadioBox) return;

					radioBox.tabIndex = -1;
					previousRadioBox.tabIndex = 0;
					previousRadioBox.focus();
					previousRadioBox.click();
				}
			});

			i++;
		}
		element.addEventListener('keydown', (e) => {
			if (e.key !== 'Enter') return;

			const checkbox = element.querySelector<HTMLInputElement>('.sui-checkbox');

			if (!checkbox) return;

			checkbox.click();
		});
	}
}

document.addEventListener('astro:page-load', loadRadioGroups);
