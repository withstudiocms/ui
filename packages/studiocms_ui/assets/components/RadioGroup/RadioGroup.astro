---
import type { StudioCMSColorway } from '../../utils/colors.js';
import { generateID } from '../../utils/generateID.js';
import './radiogroup.css';

/**
 * The props for the RadioGroup component.
 */
interface Option {
	/**
	 * The label of the option.
	 */
	label: string;
	/**
	 * The value of the option.
	 */
	value: string;
	/**
	 * Whether the option is disabled.
	 */
	disabled?: boolean;
}

/**
 * The props for the RadioGroup component.
 */
interface Props {
	/**
	 * The label of the radio group.
	 */
	label: string;
	/**
	 * The color of the radio group. Defaults to `default`.
	 */
	color?: Omit<StudioCMSColorway, 'default'>;
	/**
	 * The default value of the radio group. Needs to be one of the values in the options.
	 */
	defaultValue?: string;
	/**
	 * The options to display in the radio group.
	 */
	options: Option[];
	/**
	 * Whether the radio group is disabled. Defaults to `false`.
	 */
	disabled?: boolean;
	/**
	 * The name of the radio group.
	 */
	name?: string;
	/**
	 * Whether the radio group is required. Defaults to `false`.
	 */
	isRequired?: boolean;
	/**
	 * Whether the radio group is horizontal. Defaults to `false`.
	 */
	horizontal?: boolean;
	/**
	 * Additional classes to apply to the radio group.
	 */
	class?: string;
}

const {
	label,
	color = 'primary',
	defaultValue,
	options,
	disabled = false,
	isRequired = false,
	horizontal = false,
	name = generateID('radio'),
	class: className,
} = Astro.props;
---

<div 
  class="sui-radio-container"
  class:list={[
    disabled && "disabled",
    horizontal && "horizontal",
    color,
    className
  ]}
>
  <span>
    {label} <span class="req-star">{isRequired && "*"}</span>
  </span>
  <div class="sui-radio-inputs" role="radiogroup">
    {options.map(({ label, value, disabled: individuallyDisabled }, i) => (
      <label
        for={value}
        class="sui-radio-label"
        class:list={[ individuallyDisabled && "disabled" ]}
      >
        <div class="sui-radio-box-container">
          <div 
            class="sui-radio-box" 
            role="radio" 
            tabindex={i === 0 ? 0 : -1} 
            aria-checked={value === defaultValue}
            aria-label={label}
          />
        </div>
        <input 
          class="sui-radio-toggle"
          type="radio" 
          value={value} 
          id={value} 
          name={name} 
          checked={value === defaultValue} 
          disabled={disabled || individuallyDisabled}
          required={isRequired}
        />
        <span>{label}</span>
      </label>
    ))}
  </div>
</div>
<script>
	import 'studiocms:ui/scripts/radiogroup'
</script>
