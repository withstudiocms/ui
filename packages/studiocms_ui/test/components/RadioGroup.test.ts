import { describe, expect } from 'vitest';
import RadioGroup from '../../src/components/RadioGroup/RadioGroup.astro';
import { test } from '../fixtures/vitest/AstroContainer';

const mockProps = {
	name: 'test-radio-group',
	label: 'Test Radio Group',
	options: [
		{ label: 'Option 1', value: 'option1' },
		{ label: 'Option 2', value: 'option2' },
		{ label: 'Option 3', value: 'option3' },
	],
	defaultValue: 'option2',
};

describe('RadioGroup Component', () => {
	test('renders RadioGroup component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(RadioGroup, 'RadioGroup', { props: mockProps });
		expect(result).toMatchSnapshot();
	});

	test.for([
		{ ...mockProps, disabled: true },
		{ ...mockProps, isRequired: true },
		{ ...mockProps, horizontal: true },
	])('renders RadioGroup with props: "%s"', async (props, { renderComponent }) => {
		const result = await renderComponent(RadioGroup, 'RadioGroup', { props });
		expect(result).toMatchSnapshot();
	});
});
