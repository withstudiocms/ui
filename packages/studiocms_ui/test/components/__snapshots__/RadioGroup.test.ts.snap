// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RadioGroup Component > renders RadioGroup component correctly 1`] = `"<div class="sui-radio-container primary"> <span> Test Radio Group <span class="req-star"></span> </span> <div class="sui-radio-inputs" role="radiogroup"> <label for="option1" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="0" aria-checked="false" aria-label="Option 1"></div> </div> <input class="sui-radio-toggle" type="radio" value="option1" id="option1" name="test-radio-group"> <span>Option 1</span> </label><label for="option2" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="-1" aria-checked="true" aria-label="Option 2"></div> </div> <input class="sui-radio-toggle" type="radio" value="option2" id="option2" name="test-radio-group" checked> <span>Option 2</span> </label><label for="option3" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="-1" aria-checked="false" aria-label="Option 3"></div> </div> <input class="sui-radio-toggle" type="radio" value="option3" id="option3" name="test-radio-group"> <span>Option 3</span> </label> </div> </div> <script type="module" src="/mock/path/RadioGroup.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`RadioGroup Component > renders RadioGroup with props: "{ name: 'test-radio-group', label: 'Test Radio Group', options: [ { label: 'Option 1', value: 'option1' }, { label: 'Option 2', value: 'option2' }, { label: 'Option 3', value: 'option3' } ], defaultValue: 'option2', disabled: true }" 1`] = `"<div class="sui-radio-container disabled primary"> <span> Test Radio Group <span class="req-star"></span> </span> <div class="sui-radio-inputs" role="radiogroup"> <label for="option1" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="0" aria-checked="false" aria-label="Option 1"></div> </div> <input class="sui-radio-toggle" type="radio" value="option1" id="option1" name="test-radio-group" disabled> <span>Option 1</span> </label><label for="option2" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="-1" aria-checked="true" aria-label="Option 2"></div> </div> <input class="sui-radio-toggle" type="radio" value="option2" id="option2" name="test-radio-group" checked disabled> <span>Option 2</span> </label><label for="option3" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="-1" aria-checked="false" aria-label="Option 3"></div> </div> <input class="sui-radio-toggle" type="radio" value="option3" id="option3" name="test-radio-group" disabled> <span>Option 3</span> </label> </div> </div> <script type="module" src="/mock/path/RadioGroup.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`RadioGroup Component > renders RadioGroup with props: "{ name: 'test-radio-group', label: 'Test Radio Group', options: [ { label: 'Option 1', value: 'option1' }, { label: 'Option 2', value: 'option2' }, { label: 'Option 3', value: 'option3' } ], defaultValue: 'option2', horizontal: true }" 1`] = `"<div class="sui-radio-container horizontal primary"> <span> Test Radio Group <span class="req-star"></span> </span> <div class="sui-radio-inputs" role="radiogroup"> <label for="option1" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="0" aria-checked="false" aria-label="Option 1"></div> </div> <input class="sui-radio-toggle" type="radio" value="option1" id="option1" name="test-radio-group"> <span>Option 1</span> </label><label for="option2" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="-1" aria-checked="true" aria-label="Option 2"></div> </div> <input class="sui-radio-toggle" type="radio" value="option2" id="option2" name="test-radio-group" checked> <span>Option 2</span> </label><label for="option3" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="-1" aria-checked="false" aria-label="Option 3"></div> </div> <input class="sui-radio-toggle" type="radio" value="option3" id="option3" name="test-radio-group"> <span>Option 3</span> </label> </div> </div> <script type="module" src="/mock/path/RadioGroup.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`RadioGroup Component > renders RadioGroup with props: "{ name: 'test-radio-group', label: 'Test Radio Group', options: [ { label: 'Option 1', value: 'option1' }, { label: 'Option 2', value: 'option2' }, { label: 'Option 3', value: 'option3' } ], defaultValue: 'option2', isRequired: true }" 1`] = `"<div class="sui-radio-container primary"> <span> Test Radio Group <span class="req-star">*</span> </span> <div class="sui-radio-inputs" role="radiogroup"> <label for="option1" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="0" aria-checked="false" aria-label="Option 1"></div> </div> <input class="sui-radio-toggle" type="radio" value="option1" id="option1" name="test-radio-group" required> <span>Option 1</span> </label><label for="option2" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="-1" aria-checked="true" aria-label="Option 2"></div> </div> <input class="sui-radio-toggle" type="radio" value="option2" id="option2" name="test-radio-group" checked required> <span>Option 2</span> </label><label for="option3" class="sui-radio-label"> <div class="sui-radio-box-container"> <div class="sui-radio-box" role="radio" tabindex="-1" aria-checked="false" aria-label="Option 3"></div> </div> <input class="sui-radio-toggle" type="radio" value="option3" id="option3" name="test-radio-group" required> <span>Option 3</span> </label> </div> </div> <script type="module" src="/mock/path/RadioGroup.astro?astro&type=script&index=0&lang.ts"></script>"`;
